pkgs/by-name/de/deltatouch/package.nix

```diff
@@ -0,0 +1,93 @@
+{ lib
+, stdenv
+, fetchFromGitea
+, cmake
+, intltool
+, libdeltachat
+, lomiri
+, qt5
+, quirc
+}:
+
+stdenv.mkDerivation {
+  pname = "deltatouch";
+  version = "1.3.1-unstable-2024-02-25";
+
+  src = fetchFromGitea {
+    domain = "codeberg.org";
+    owner = "lk108";
+    repo = "deltatouch";
+    rev = "dd86a77291a6e9f1c363e7cdf5999e58fb5b84e5";
+    hash = "sha256-03J2YLXF1uDnTmDsrxWjlfDsN7mcc68MbWy/HTIcM2E=";
+    fetchSubmodules = true;
+  };
+
+  nativeBuildInputs = [
+    qt5.wrapQtAppsHook
+    intltool
+    cmake
+  ];
+
+  buildInputs = [
+    qt5.qtbase
+    qt5.qtwebengine
+    qt5.qtquickcontrols2
+    lomiri.lomiri-ui-toolkit
+    lomiri.lomiri-ui-extras
+    lomiri.lomiri-api
+    lomiri.lomiri-indicator-network # Lomiri.Connectivity module
+    lomiri.qqc2-suru-style
+  ];
+
+  postPatch = ''
+    # Fix all sorts of install locations
+    substituteInPlace CMakeLists.txt \
+      --replace-fail 'set(DATA_DIR /)' "set(DATA_DIR $out/share)" \
+      --replace-fail 'RUNTIME DESTINATION ''${CMAKE_INSTALL_PREFIX}' 'RUNTIME DESTINATION ''${CMAKE_INSTALL_PREFIX}/bin' \
+      --replace-fail 'assets/logo.svg DESTINATION assets' 'assets/logo.svg DESTINATION ''${CMAKE_INSTALL_PREFIX}/share/icons/hicolor/scalable/apps RENAME deltatouch.svg' \
+      --replace-fail "\''${DESKTOP_FILE_NAME} DESTINATION \''${DATA_DIR}" "\''${DESKTOP_FILE_NAME} DESTINATION \''${CMAKE_INSTALL_PREFIX}/share/applications"
+
+    substituteInPlace plugins/DeltaHandler/CMakeLists.txt plugins/DTWebEngineProfile/CMakeLists.txt \
+      --replace-fail 'set(QT_IMPORTS_DIR "/lib/''${ARCH_TRIPLET}")' 'set(QT_IMPORTS_DIR "${placeholder "out"}/${qt5.qtbase.qtQmlPrefix}")'
+
+    substituteInPlace po/CMakeLists.txt \
+      --replace-fail 'INSTALL_DIR ''${CMAKE_INSTALL_LOCALEDIR}/share/locale' 'INSTALL_DIR ''${CMAKE_INSTALL_LOCALEDIR}'
+
+    # Fix import of library dependencies
+    substituteInPlace plugins/DeltaHandler/CMakeLists.txt \
+      --replace-fail 'IMPORTED_LOCATION "''${CMAKE_CURRENT_BINARY_DIR}/libdeltachat.so"' 'IMPORTED_LOCATION "${lib.getLib libdeltachat}/lib/libdeltachat.so"' \
+      --replace-fail 'IMPORTED_LOCATION "''${CMAKE_CURRENT_BINARY_DIR}/libquirc.so.1.2"' 'IMPORTED_LOCATION "${lib.getLib quirc}/lib/libquirc.so.1.0"'
+
+    # Fix icon reference in desktop file
+    substituteInPlace deltatouch.desktop.in \
+      --replace-fail 'Icon=assets/logo.svg' 'Icon=deltatouch'
+
+    # Make localisation work
+    substituteInPlace qml/Main.qml \
+      --replace-fail 'Component.onCompleted: {' 'Component.onCompleted: {
+        i18n.domain = "deltatouch.lotharketterer"
+        i18n.bindtextdomain("deltatouch.lotharketterer", "${placeholder "out"}/share/locale")'
```

Comment:
This could prolly still be done better… Either submitted upstream somehow (and tested in the click usecase) as mentioned earlier, or done as a vendored patch (which would require occasional fixing on bumps, and maybe some shuffling around to get the correct `$out` in there)…

Prolly fine for now, but I'll put giving this another look onto my Lomiri TODO list if you can't take a look at tackling this.

Reply:
I can try to add it to Main.qml, but I'm not sure how to handle the "placeholder out" part.

Reply:
Some Lomiri stuff does it by via a context property, which sets it via a define in C/C++ land:

https://gitlab.com/ubports/development/core/lomiri-system-settings/-/blob/e21e51a333f8f285f97953abc03b970df149fadf/src/qml/MainWindow.qml#L77
```qml
    Component.onCompleted: {
        i18n.domain = "lomiri-system-settings"
        i18n.bindtextdomain("lomiri-system-settings", i18nDirectory)
```
https://gitlab.com/ubports/development/core/lomiri-system-settings/-/blob/e21e51a333f8f285f97953abc03b970df149fadf/src/main.cpp#L86
```c++
    view.rootContext()->setContextProperty("i18nDirectory", mountPoint + I18N_DIRECTORY);
```

https://gitlab.com/ubports/development/core/lomiri-system-settings/-/blob/e21e51a333f8f285f97953abc03b970df149fadf/src/CMakeLists.txt#L3
```cmake
add_definitions(-DI18N_DIRECTORY="${CMAKE_INSTALL_PREFIX}/share/locale")
```
(This should be `CMAKE_INSTALL_FULL_LOCALEDIR`, fix has been submitted to them.)

---

Alternatively, you could turn the file into a template:
```qml
    Component.onCompleted: {
        i18n.domain = "deltatouch.lotharketterer"
        i18n.bindtextdomain("deltatouch.lotharketterer", "@CMAKE_INSTALL_FULL_LOCALEDIR@")
```
and have CMake replace the placeholder at build time:
```cmake
configure_file(${CMAKE_CURRENT_SOURCE_DIR}/Main.qml.in ${CMAKE_CURRENT_BINARY_DIR}/Main.qml @ONLY)
```

---

Either method would also require using
```cmake
include(GNUInstallDirs)
```
to have proper guarantees on `CMAKE_INSTALL_FULL_LOCALEDIR`, which also opens up further fix upstreaming for install locations from the `postPatch` script.

Reply:
Thanks, I'll have a look

Reply:
With the first option you suggested, it works on Ubuntu Touch, but not on Nix. [Here's](https://codeberg.org/lk108/deltatouch/commit/b650643ae90a72a3d699eaf374094d279e8d7972) the corresponding commit.

On Nix, the define in C++ as set by cmake is `/nix/store/2az7rny9qv110fk57paxqwvpy7hj8l3l-deltatouch-1.4.0-unstable-2024-03-08/share/locale`, which actually contains the lang dirs like `de/LC_MESSAGES/deltatouch.lotharketterer.mo`. Yet the translations are not used.

Reply:
> it works on Ubuntu Touch, but not on Nix

Seems to work fine for me. Not sure what state the application is stuck in on my end, so I had to add a translation for it to show up (immediately jumps to this "Known accounts" screen that I can't exit), but the catalogue gets loaded & can be used with that commit.

![image](https://github.com/NixOS/nixpkgs/assets/23431373/96daa255-7757-420d-9a8a-917ea43e1ad1)

Reply:
It's merged into main, commit dcfdd8a0fca5fff10d0383f77f4c0cbea302de00. In package.nix, the `substituteInPlace po/CMakeLists.txt`and `substituteInPlace qml/Main.qml` parts should now be removed, right?

For some reason, for me on Nix, `QLocale::system().name()` in C++ returns de_DE, but `i18n.language` in Main.qml is empty, so it doesn't even try to open any .mo file.

Reply:
> You don't have the icons set up

Gah, right. Laziness be damned… :sweat_smile: 

> In package.nix, the `substituteInPlace po/CMakeLists.txt` and `substituteInPlace qml/Main.qml` parts should now be removed, right?

Yes.

> `QLocale::system().name()` in C++ returns de_DE, but `i18n.language` in Main.qml is empty.

Hmm. You might need to look into how LUITK sets this, and what environment variables are used to determine it: https://gitlab.com/ubports/development/core/lomiri-ui-toolkit/-/blob/355b2c09a64efddc80b7ea9fbcc9d1462d7573cf/src/LomiriToolkit/i18n.cpp#L62-71

Reply:
> Hmm. You might need to look into how LUITK sets this, and what environment variables are used to determine it

Thanks for the link. According to the comments in i18n.cpp, it should work with my env. For the moment, I am out of ideas. Do you consider it a blocker if this is not resolved?

Reply:
Bummer. :slightly_frowning_face: 

I think we should at least rewrite this substitution here as a vendored patch then, due to its insertion of multiple lines.

Reply:
Well.. now translations don't show up with the substitution either. Something must be wrong with my setup.

Reply:
FWIW, applying https://codeberg.org/lk108/deltatouch/commit/dcfdd8a0fca5fff10d0383f77f4c0cbea302de00 worksfine for me. Might be a setup issue then?

![image](https://github.com/NixOS/nixpkgs/assets/23431373/6c6fc90a-9805-487b-8f66-a22010776e6c)

I'll send a suggestion to fetch that as a patch & drop/adjust the substitutions affected by it, then this LGTM.

Reply:
> I'll send a suggestion to fetch that as a patch & drop/adjust the substitutions affected by it, then this LGTM. That fine with you?

Not sure if you're asking me, but I'm fine with it :-)